### PR TITLE
[Colossal-Llama-2] Add finetuning Colossal-Llama-2 example

### PR DESCRIPTION
**Pull request by chengeharrison**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Add an example of finetuning Colossal-Llama-2.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/Colossal-LLaMA-2/train_sft.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
+#!/usr/bin/env python3
```
Comment (TongLi3701) [inline applications/Colossal-LLaMA-2/train_sft.py:1]: It seems that this file is the same as `train.py`.

If there is not difference, we can remove this file.

**Review by TongLi3701 — COMMENTED** [review]
Thanks Yuanchen. Overall looks good. Some comments: 

1. Please avoid repetitive code as much as possible.
2. If some of the function name has been changed, you should change it in the corresponding files as well.
3. Please also update the README file for model fine-tuning.

**Comment by ver217** [comment]
Could you update Coati based on this PR? Thanks!

File: applications/Colossal-LLaMA-2/train_sft.py (hunk 1 above)
Comment (chengeharrison) [reply applications/Colossal-LLaMA-2/train_sft.py:1]: Add gradient accumulation here.

File: applications/Colossal-LLaMA-2/train_sft.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+# -*- coding: utf-8 -*-
+"""
+Continual Pre-training of LLaMA-2 developed by Colossal-AI Team
```
Comment (Camille7777) [inline applications/Colossal-LLaMA-2/train_sft.py:4]: Here should be "Supervised fine-tuning of Colossal-LLaMA-2-base"?

File: applications/Colossal-LLaMA-2/train_sft.example.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    "PATH TO THE DATASET"
+)
+
+TIMESTAMP=$(date +%Y-%m-%d-%H-%M-%S)
+FULL_PROJECT_NAME="${PROJECT_NAME}-${TIMESTAMP}"
+SAVE_DIR="${PARENT_SAVE_DIR}${FULL_PROJECT_NAME}"
+TENSORBOARD_DIR="${PARENT_TENSORBOARD_DIR}${FULL_PROJECT_NAME}"
+CONFIG_FILE="${PARENT_CONFIG_FILE}${FULL_PROJECT_NAME}.json"
+
+colossalai run --nproc_per_node 8 --hostfile hostfile --master_port 30013 train.py \
```
Comment (Camille7777) [inline applications/Colossal-LLaMA-2/train_sft.example.sh:28]: train_sft.py

**Comment by TongLi3701** [comment]
> Could you update Coati based on this PR? Thanks!

Coati will be a separate project focusing on general model fine-tuning and RLHF.

File: applications/Colossal-LLaMA-2/colossal_llama2/dataset/spliced_and_tokenized_dataset.py [hunk 4]
```diff
@@ -62,6 +69,126 @@ def supervised_tokenize(
...
+        )
+
+    target_turn = turns[target_turn_index - 1]
+    prompt = template.get_prompt(2 * target_turn)
+    tokenized = tokenizer([prompt], add_special_tokens=False)["input_ids"][0]
+
+    # Uncomment the following to check whether `bisect_right` is right.
+    # if 2 * target_turn < len(template.messages):
+    #     length_to_next_turn = len(tokenizer([template.get_prompt(2*target_turn+2)], add_special_tokens=False)["input_ids"][0])
+    #     assert length_to_next_turn > max_length - 1, logger.info(f"The length of the prompt until the next turn after tokenization is {length_to_next_turn}, which is smaller than {max_length - 1}")
```
Comment (TongLi3701) [inline applications/Colossal-LLaMA-2/colossal_llama2/dataset/spliced_and_tokenized_dataset.py:139]: Debug code can be removed.

**Review by TongLi3701 — CHANGES_REQUESTED** [review]
Thanks Yuanchen.

Please also update readme for llama2 sft.

File: applications/Colossal-LLaMA-2/colossal_llama2/utils/neftune_patch.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+def deactivate_neftune(model, neftune_hook_handle):
+    """
+    Deactivates the neftune method. Make sure to call `_activate_neftune` first.
+    """
+    unwrapped_model = unwrap(model)
+    embeddings = unwrapped_model.get_input_embeddings()
+
+    neftune_hook_handle.remove()
+    del embeddings.neftune_noise_alpha, unwrapped_model
```
Comment (TongLi3701) [inline applications/Colossal-LLaMA-2/colossal_llama2/utils/neftune_patch.py:70]: Why is it necessary to delete these two here? 

https://github.com/ludwig-ai/ludwig/pull/3744/files#diff-68b4ac2f32ed3513a319b21adc40f844ca4ef80c4fe9b9556fe610a16e3976f4R304-R345

In this PR ,there are some tests for `activate` and `deactivate`. It is better to add tests for this method

Comment (Camille7777) [reply applications/Colossal-LLaMA-2/colossal_llama2/utils/neftune_patch.py:70]: @TongLi3701 We just del the neftune_noise_alpha attribute here, refer to HF Transformer: https://github.com/huggingface/transformers/blob/083e36923a19650fa264c4173db2f63ab124bb27/src/transformers/trainer.py#L634-L649

File: applications/Colossal-LLaMA-2/colossal_llama2/utils/neftune_patch.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    return model, neftune_hook_handle
+
+
+def deactivate_neftune(model, neftune_hook_handle):
+    """
+    Deactivates the neftune method. Make sure to call `_activate_neftune` first.
+    """
+    unwrapped_model = unwrap(model)
+    embeddings = unwrapped_model.get_input_embeddings()
```
Comment (Camille7777) [inline applications/Colossal-LLaMA-2/colossal_llama2/utils/neftune_patch.py:67]: @chengeharrison Can we just implement `deactivate_neftune` in line with `activate_neftune` as following:
`embeddings = unwrap(model).get_input_embeddings()`
and remove the unwrapped_model in del part？

File: applications/Colossal-LLaMA-2/colossal_llama2/utils/neftune_patch.py (hunk 5 above)
Comment (TongLi3701) [reply applications/Colossal-LLaMA-2/colossal_llama2/utils/neftune_patch.py:70]: I see. Thanks.

File: applications/Colossal-LLaMA-2/inference_example.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
+
+import os
+import torch
+import argparse
+from colossalai.logging import get_dist_logger
+from transformers import AutoTokenizer, AutoModelForCausalLM
```
Comment (TongLi3701) [inline applications/Colossal-LLaMA-2/inference_example.py:7]: Imports should be reorganised.

**Review by TongLi3701 — APPROVED** [review]
Thanks Yuting.

Some minor comments. Please address the comments and merge.

**Review by Camille7777 — APPROVED** [review]
